lib/ui/painting/image_generator_apng.cc

```diff
@@ -0,0 +1,616 @@
+
+    void Premultiply() {
+      for (int i = 0; i < 3; i++) {
+        channel[i] = channel[i] * GetAlpha() / 0xFF;
+      }
+    }
+
+    void Unpremultiply() {
+      if (GetAlpha() == 0) {
+        channel[0] = channel[1] = channel[2] = 0;
+        return;
+      }
+      for (int i = 0; i < 3; i++) {
+        channel[i] = channel[i] * 0xFF / GetAlpha();
+      }
+    }
+  };
+
+  FML_DCHECK(frame_info.bytesPerPixel() == sizeof(Pixel));
+
+  for (int y = 0; y < frame_info.height(); y++) {
+    auto src_row = frame.pixels.data() + y * frame_row_bytes;
+    auto dst_row = static_cast<uint8_t*>(pixels) +
+                   (y + frame.y_offset) * row_bytes +
+                   frame.x_offset * frame_info.bytesPerPixel();
+
+    switch (frame.frame_info->blend_mode) {
+      case SkCodecAnimation::Blend::kSrcOver: {
+        for (int x = 0; x < frame_info.width(); x++) {
+          auto x_offset_bytes = x * frame_info.bytesPerPixel();
+
+          Pixel src = *reinterpret_cast<Pixel*>(src_row + x_offset_bytes);
+          Pixel* dst_p = reinterpret_cast<Pixel*>(dst_row + x_offset_bytes);
+          Pixel dst = *dst_p;
+
+          // Ensure both colors are premultiplied for the blending operation.
+          if (info.alphaType() == kUnpremul_SkAlphaType) {
+            dst.Premultiply();
+          }
+          if (frame_info.alphaType() == kUnpremul_SkAlphaType) {
+            src.Premultiply();
+          }
+
+          for (int i = 0; i < 4; i++) {
+            dst.channel[i] = src.channel[i] +
+                             dst.channel[i] * (0xFF - src.GetAlpha()) / 0xFF;
+          }
+
+          // The final color is premultiplied. Unpremultiply to match the
+          // backdrop surface if necessary.
+          if (info.alphaType() == kUnpremul_SkAlphaType) {
+            dst.Unpremultiply();
+          }
+
+          *dst_p = dst;
+        }
+        break;
+      }
+      case SkCodecAnimation::Blend::kSrc:
+        memcpy(dst_row, src_row, frame_row_bytes);
+        break;
+    }
+  }
+
+  return true;
+}
+
+std::unique_ptr<ImageGenerator> APNGImageGenerator::MakeFromData(
+    sk_sp<SkData> data) {
+  // Ensure the buffer is large enough to at least contain the PNG signature
+  // and a chunk header.
+  if (data->size() < 8 + sizeof(ChunkHeader)) {
+    return nullptr;
+  }
+  // Validate the full PNG signature.
+  const uint8_t* data_p = static_cast<const uint8_t*>(data.get()->data());
+  if (png_sig_cmp(static_cast<png_const_bytep>(data_p), 0, 8)) {
+    return nullptr;
+  }
+
+  // Validate the header chunk.
+  const ChunkHeader* chunk = reinterpret_cast<const ChunkHeader*>(data_p + 8);
+  if (!IsValidChunkHeader(data_p, data->size(), chunk) ||
+      chunk->get_data_length() != sizeof(ImageHeaderChunkData) ||
+      chunk->get_type() != kImageHeaderChunkType) {
+    return nullptr;
+  }
+
+  // Walk the chunks to find the "animation control" chunk. If an "image data"
+  // chunk is found first, this PNG is not animated.
+  do {
```

Comment:
Move the GetNextChunk call to the start of the loop.  The initial chunk is known to be an image header chunk and does not need to be checked.

Perhaps something like:
```
while (true) {
  chunk = GetNextChunk();
  if (chunk == nullptr)
    return nullptr;
  if (chunk->get_type() == ImageData)
    return nullptr;
  if (chunk->get_type() == AnimationControl)
    break;
}